src/connector/src/sink/sqlserver.rs

```diff
@@ -0,0 +1,625 @@
+}
+
+pub struct SqlServerSinkWriter {
+    config: SqlServerConfig,
+    schema: Schema,
+    pk_indices: Vec<usize>,
+    is_append_only: bool,
+    sql_client: SqlClient,
+    ops: Vec<SqlOp>,
+}
+
+impl SqlServerSinkWriter {
+    pub async fn new(
+        config: SqlServerConfig,
+        schema: Schema,
+        pk_indices: Vec<usize>,
+        is_append_only: bool,
+    ) -> Result<Self> {
+        let sql_client = SqlClient::new(&config).await?;
+        let writer = Self {
+            config,
+            schema,
+            pk_indices,
+            is_append_only,
+            sql_client,
+            ops: vec![],
+        };
+        Ok(writer)
+    }
+
+    fn delete_one(&mut self, row: RowRef<'_>) {
+        // TODO: avoid into_owned_row
+        self.ops.push(SqlOp::Delete(row.into_owned_row()));
+    }
+
+    fn upsert_one(&mut self, row: RowRef<'_>) {
+        // TODO: avoid into_owned_row
+        if self.is_append_only {
+            self.ops.push(SqlOp::Insert(row.into_owned_row()));
+        } else {
+            self.ops.push(SqlOp::Merge(row.into_owned_row()));
+        }
+    }
+
+    async fn flush(&mut self) -> Result<()> {
+        use std::fmt::Write;
+        if self.ops.is_empty() {
+            return Ok(());
+        }
+        let mut query_str = String::new();
+        let col_num = self.schema.fields.len();
+        let mut next_param_id = 1;
+        let non_pk_col_indices = (0..col_num)
+            .filter(|idx| !self.pk_indices.contains(idx))
+            .collect::<Vec<usize>>();
+        let all_col_names = self
+            .schema
+            .fields
+            .iter()
+            .map(|f| f.name.as_str())
+            .collect::<Vec<_>>()
+            .join(",");
+        let all_source_col_names = self
+            .schema
+            .fields
+            .iter()
+            .map(|f| format!("SOURCE.{}", f.name))
+            .collect::<Vec<_>>()
+            .join(",");
+        let pk_match = self
+            .pk_indices
+            .iter()
+            .map(|idx| {
+                format!(
+                    "SOURCE.{}=TARGET.{}",
+                    &self.schema[*idx].name, &self.schema[*idx].name
+                )
+            })
+            .collect::<Vec<_>>()
+            .join(" AND ");
+        let param_placeholders = |param_id: &mut usize| {
+            let params = (*param_id..(*param_id + col_num))
+                .map(|i| format!("@P{}", i))
+                .collect::<Vec<_>>()
+                .join(",");
+            *param_id += col_num;
+            params
+        };
+        let set_all_source_col = non_pk_col_indices
+            .iter()
+            .map(|idx| {
+                format!(
+                    "{}=SOURCE.{}",
+                    &self.schema[*idx].name, &self.schema[*idx].name
+                )
+            })
+            .collect::<Vec<_>>()
+            .join(",");
+        // TODO: avoid repeating the SQL
+        for op in &self.ops {
```

Comment:
It seems we generate one SQL command for each `op` here. We should batch ops of a chunk in one SQL, for example a `INSERT INTO t VALUES (1,1), (1,2), ()..` to insert multiple rows in one sql.

Reply:
The problem here is this will break the order of upstream rows. e.g. stream chunk may contains a insert followed by a delete then another insert. If we batch these two insert, the order becomes different .

Reply:
Currently we only do the compaction when the pk and stream key mismatch and it is not for any cases. And for this optimization I guess we can do it for a consecutive sequence of identical operations. I think it is a general optimization for any downstream system, and we need some benchmarks to verify its effectiveness.

Reply:
> Currently we only do the compaction when the pk and stream key mismatch and it is not for any cases

Hmm, but I have applied this assumption to the jdbc sink for a long time. 😅
https://github.com/risingwavelabs/risingwave/blob/8d16088c244801df1bd2e6b1cc06516d817756b6/java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java#L326-L328

Reply:
This could be a issue, but it's hard to trigger... The first condition is that it only happens when there is a mismatch between the stream key and the primary key. The second is that: Because by advancing the delete operation here, issues will only happend when such a sequence appears within a chunk (for the same primary key), where `Insert, Delete` leads to the expected deletion not taking effect. Note that even if the sequence is `Insert, Delete, Insert`, there won't be a problem, because there is no disorder among the changes with the same operation. In this case, the final effective operation is still the last insert operation.
I am not sure if we can do the compaction in any cases... Currently, it buffered all changes in one epoch, leading to reduce latency of the sink. But we can do this compaction within a single chunk, which might take more overhead.

Reply:
https://github.com/risingwavelabs/risingwave/pull/17174